_posts/2023-01-07-datafusion-16.0.0.md

```diff
@@ -0,0 +1,308 @@
+
+* [Greptime DB](https://github.com/GreptimeTeam/greptimedb)
+* [Synnada](https://synnada.ai/)
+* [PRQL](https://github.com/PRQL/prql-query)
+- [Parseable](https://github.com/parseablehq/parseable)
+* [SeaFowl](https://github.com/splitgraph/seafowl)
+
+
+## Performance 🚀
+
+Performance and efficiency are core values for
+DataFusion. While there is still a gap between DataFusion and the best of
+breed, tightly integrated systems such as [DuckDB](https://duckdb.org)
+and [Polars](https://www.pola.rs/), DataFusion is
+closing the gap quickly. Performance highlights from the last three
+months:
+
+* Up to 30% Faster Sorting and Merging using the new [Row Format](https://arrow.apache.org/blog/2022/11/07/multi-column-sorts-in-arrow-rust-part-1/)
+* [Advanced predicate pushdown](https://arrow.apache.org/blog/2022/12/26/querying-parquet-with-millisecond-latency/), directly on parquet, directly from object storage, enabling sub millisecond filtering. <!-- Andrew nots: we should really get this turned on by default -->
+* `70%` faster `IN` expressions evaluation ([#4057])
+* Sort and partition aware optimizations ([#3969] and  [#4691])
+* Filter selectivity analysis ([#3868])
+
+## Runtime Resource Limits
+
+Previously, DataFusion could potentially use unbounded amounts of memory for certain queries that included Sorts, Grouping or Joins.
+
+In version 16.0.0, it is possible to limit DataFusion's memory usage for Sorting and Grouping. We are looking for help adding similar limiting for Joins as well as expanding our algorithms to optionally spill to secondary storage. See [#3941] for more detail.
+
+
+## SQL Window Functions
+
+[SQL Window Functions](https://en.wikipedia.org/wiki/Window_function_(SQL)) are useful for a variety of analysis and DataFusion's implementation support expanded significantly:
+
+- Custom window frames such as `... OVER (ORDER BY ... RANGE BETWEEN 0.2 PRECEDING AND 0.2 FOLLOWING)`
+- Unbounded window frames such as `... OVER (ORDER BY ... RANGE UNBOUNDED ROWS PRECEDING)`
+- Support for the `NTILE` window function ([#4676])
+- Support for `GROUPS` mode ([#4155])
+
+
+# Improved Joins
+
+Joins are often the most complicated operations to handle well in
+analytics systems and DataFusion 16.0.0 offers significant improvements
+such as
+
+- Cost based optimizer (CBO) automatically reorders join evaluations, selects algorithms (Merge / Hash), and pick build side based on available statistics and join type (`INNER`, `LEFT`, etc) ([#4219])
+- Fast non `column=column` equijoins such as `JOIN ON a.x + 5 = b.y`
+- Better performance on non-equijoins ([#4562]) <!-- TODO is this a good thing to mention as any time this is usd the query is going to go slow or the data size is small -->
+
+# Streaming Execution
+
+One emerging use case for Datafusion is as a foundation for
+streaming-first data platforms. An important prerequisite
+is support for incremental execution if queries that can be computed
```

Comment:
Grammar nit: 'if queries that can' -> 'for queries that can ...'

Reply:
```suggestion
is support for incremental execution for queries that can be computed
```